File: packages/java/src/main/java/com/digitalt3/commons/api/Tracer.java
package com.digitalt3.commons.api;

import java.util.Map;
import java.util.function.Consumer;
import java.util.function.Function;

/**
 * Lightweight tracer that creates W3C-compatible spans via scoped {@link LogContext}.
 *
 * @since 0.1.0
 */
public interface Tracer {

    /**
     * Start a new span as a child of the currently active span when present.
     *
     * @param name span name used for completion events
     * @return an open span that must be ended or closed
     */
    Span startSpan(String name);

    /**
     * Start a new span with optional attributes applied to the span context.
     *
     * @param name span name used for completion events
     * @param attributes optional attributes attached to span completion events
     * @return an open span that must be ended or closed
     */
    Span startSpan(String name, Map<String, Object> attributes);

    /**
     * Run work under an active span and end the span when the callback returns.
     *
     * @param name span name
     * @param work callback receiving the active span
     * @param <T> result type
     * @return the callback result
     */
    <T> T withSpan(String name, Function<Span, T> work);

    /**
     * Run work under an active span and end the span when the callback returns.
     *
     * @param name span name
     * @param work callback receiving the active span
     */
    void withSpan(String name, Consumer<Span> work);
}

File: packages/node/src/api/Tracer.ts
import { Attributes } from './types';

/**
 * Options supplied when starting a span.
 */
export interface SpanOptions {
  /** Optional attributes recorded on the span completion event. */
  attributes?: Attributes;
}

/**
 * Lightweight W3C-compatible span handle.
 *
 * Spans participate in logging via scoped `LogContext` (`trace.id`, `span.id`,
 * `parent.span.id`) rather than requiring the full OpenTelemetry SDK.
 */
export interface Span {
  /** Human-readable span name. */
  readonly name: string;

  /** W3C trace id (32 lowercase hex characters). */
  readonly traceId: string;

  /** W3C span id (16 lowercase hex characters). */
  readonly spanId: string;

  /** Parent span id when this span is nested. */
  readonly parentSpanId?: string;

  /**
   * End the span, restore the previous log context, and optionally emit a
   * span-completion LogEvent.
   */
  end(): void;

  /**
   * Emit a LogEvent under the active span context.
   *
   * @param name - UPPER_SNAKE_CASE event name, or a free-form label mapped to `SPAN_EVENT`.
   * @param attributes - Optional event fields merged into the emitted event.
   */
  addEvent(name: string, attributes?: Attributes): void;
}

/**
 * Creates and scopes W3C-compatible spans for structured logging.
 */
export interface Tracer {
  /**
   * Start a span and activate its trace context for the current async scope.
   *
   * @param name - Span name used in completion events.
   * @param options - Optional span attributes.
   * @returns An active span that must be ended by the caller.
   */
  startSpan(name: string, options?: SpanOptions): Span;

  /**
   * Run a callback with an active span scope and end the span on exit.
   *
   * @param name - Span name used in completion events.
   * @param callback - Work executed while the span context is active.
   * @param options - Optional span attributes.
   * @returns The callback result.
   */
  withSpan<T>(name: string, callback: (span: Span) => T, options?: SpanOptions): T;
}
